op-supervisor/supervisor/backend/db/db.go

```diff
@@ -0,0 +1,494 @@
+	if err := db.init(); err != nil {
+		return fmt.Errorf("failed to find new last entry context: %w", err)
+	}
+	return nil
+}
+
+// writeSearchCheckpoint appends search checkpoint and canonical hash entry to the log
+// type 0: "search checkpoint" <type><uint64 block number: 8 bytes><uint32 event index offset: 4 bytes><uint64 timestamp: 8 bytes> = 20 bytes
+// type 1: "canonical hash" <type><parent blockhash truncated: 20 bytes> = 21 bytes
+func (db *DB) writeSearchCheckpoint(blockNum uint64, logIdx uint32, timestamp uint64, blockHash common.Hash) error {
+	var entry [entrySize]byte
+	entry[0] = typeSearchCheckpoint
+	binary.LittleEndian.PutUint64(entry[1:9], blockNum)
+	binary.LittleEndian.PutUint32(entry[9:13], logIdx)
+	binary.LittleEndian.PutUint64(entry[13:21], timestamp)
+	if err := db.writeEntry(entry); err != nil {
+		return err
+	}
+	return db.writeCanonicalHash(blockHash)
+}
+
+func (db *DB) readSearchCheckpoint(entryIdx int64) (checkpointData, error) {
+	data, err := db.readEntry(entryIdx)
+	if err != nil {
+		return checkpointData{}, fmt.Errorf("failed to read entry %v: %w", entryIdx, err)
+	}
+	if data[0] != typeSearchCheckpoint {
+		return checkpointData{}, fmt.Errorf("%w: expected search checkpoint at entry %v but was type %v", ErrDataCorruption, entryIdx, data[0])
+	}
+	return db.parseSearchCheckpoint(data), nil
+}
+
+func (db *DB) parseSearchCheckpoint(data [entrySize]byte) checkpointData {
+	return checkpointData{
+		blockNum:  binary.LittleEndian.Uint64(data[1:9]),
+		logIdx:    binary.LittleEndian.Uint32(data[9:13]),
+		timestamp: binary.LittleEndian.Uint64(data[13:21]),
+	}
+}
+
+// writeCanonicalHash appends a canonical hash entry to the log
+// type 1: "canonical hash" <type><parent blockhash truncated: 20 bytes> = 21 bytes
+func (db *DB) writeCanonicalHash(blockHash common.Hash) error {
+	var entry [entrySize]byte
+	entry[0] = typeCanonicalHash
+	copy(entry[1:21], truncateHash(blockHash))
+	return db.writeEntry(entry)
+}
+
+func (db *DB) readCanonicalHash(entryIdx int64) (TruncatedHash, error) {
+	data, err := db.readEntry(entryIdx)
+	if err != nil {
+		return TruncatedHash{}, fmt.Errorf("failed to read entry %v: %w", entryIdx, err)
+	}
+	if data[0] != typeCanonicalHash {
+		return TruncatedHash{}, fmt.Errorf("%w: expected canonical hash at entry %v but was type %v", ErrDataCorruption, entryIdx, data[0])
+	}
+	return db.parseCanonicalHash(data), nil
+}
+
+func (db *DB) parseCanonicalHash(data [24]byte) TruncatedHash {
+	return data[1:21]
+}
+
+// writeInitiatingEvent appends an initiating event to the log
+// type 2: "initiating event" <type><blocknum diff: 1 byte><event flags: 1 byte><event-hash: 20 bytes> = 23 bytes
+func (db *DB) writeInitiatingEvent(postState logContext, log common.Hash) error {
+	var entry [entrySize]byte
+	entry[0] = typeInitiatingEvent
+	blockDiff := postState.blockNum - db.lastEntryContext.blockNum
+	if blockDiff > math.MaxUint8 {
+		// TODO(optimism#10857): Need to find a way to support this.
```

Comment:
If this happens, the DB thus far is invalid right? Why the TODO? Returning an error should be sufficient, if we can never get into this state with public-api usage of the DB.

Reply:
No, this will happen anytime you get a chain where there's 255 blocks or more without a log.  That could easily happen if the sequencing window caused a reorg since there are often long stretches without any deposit transactions that emit events.  I suspect the answer will be to introduce a new event type that supports just increasing/setting the block number - most of the time it won't be used so taking 24 bytes to skip over large empty block ranges seems like a good trade off.